### DISCUSSION TITLE
Question regarding $count in SomeOdataOpenapiExample

### DISCUSSION THREAD

**Discussion by HOGENTJasperDeSmet** [dotnet/aspnet-api-versioning#938] [Q&A] [answered]
Hello,

I was testing out the `SomeOdataOpenApiExample` and trying out the different parameters. And was wondering why the count operator doesn't seem to return the total count of items.  All the other parameters seem to work fine.

Sorry, If there is an obvious answer as to why this isn't the case,  I wasn't able to find it.

Thanks in advance

**Comment by commonsensesoftware** [comment, marked as answer]
This is a great question. I believe you are asking why you don't get the `@odata.count` annotation in the response when `$count=true`. Correct me if I'm mistaken. This happens because you are only using _some_ of OData. Specifically, you are only using the OData query options, you aren't using any other part of OData. In order to get an OData formatted response, you need the OData media type formatters which are part of the full OData stack. I'm not sure if you can use the OData media type formatters without an EDM; I suspect not. Even if they are supported, you should be aware that it will likely change the shape of your response.

If your response is:

```json
[
 {...},
 {...}
]
```

The OData equivalent with `$count` would be:

```json
{
   "@count": 42,
   "value": [
      {...},
      {...}
   ]
}
```

You have a couple of options. If you're ok with the OData-style response, you can make it _look_ like that without actually using OData. You can wrap your response with something like:

```c#
public class ODataValue<T>
{
    public ODataValue( IEnumerable<T> value, long count )
    {
        Value = value;
        Count = count;
    }

    [JsonProperty( "@count") ]
    public long Count { get; }

    [JsonProperty( "value" )]
    public IEnumerable<T> Value { get; }
}
```

And then just wrap your response:

```c#
return Ok(new ODataValue<Order>(orders, totalCount));
```

If you **don't** want to use the OData format, then you have a couple of other options. If you care about REST and following what HTTP already defines, then you can return `206` with the [Content-Range](https://datatracker.ietf.org/doc/html/rfc7233#section-4-2) header. You _could_ interpret the presence of `$skip`, `$top`, and/or `$count` to imply `Accept-Ranges: items` from the client. `items` is a custom unit type, but it makes sense. For example:

```http
GET /orders?$count=true HTTP/2
Host: localhost
```

The server can enforce server-side paging, but still return a **count** with:

```http
HTTP/2 206 Partial Content
Content-Range: items 0-24/100
Content-Type: application/json
Content-Length: 420

[{...}, {...}, {...}]
```

This would indicate to the client that they are receiving a partial response; specifically, the first 25 items (it's zero-based) out of 100. I suspect the range being returned _should_ be known, but if it's not, you can also return `Content-Type: items */100` to indicate just the count.

Failing any of that, you can use your own custom HTTP header, but I think you'd need a compelling reason to do so. There's stuff already defined for that purpose.

**Reply by HOGENTJasperDeSmet** [reply]
Wow thanks for the quick reply and the comprehensive answer. This was exactly what I was looking for.  I'll  probably just wrap the response since the client is already adapted to use Odata,

 Thanks! 😊
